src/serialization/serialization.rs

```diff
@@ -0,0 +1,412 @@
+extern crate byteorder;
+
+use super::{Counter, Histogram};
+use std::io::{self, Cursor, Read, Write};
+use std;
+use super::num::ToPrimitive;
+use self::byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+
+#[path = "tests.rs"]
+#[cfg(test)]
+mod tests;
+
+#[path = "benchmarks.rs"]
+#[cfg(all(test, feature = "bench_private"))]
+mod benchmarks;
+
+const V2_COOKIE_BASE: u32 = 0x1c849303;
+
+const V2_COOKIE: u32 = V2_COOKIE_BASE | 0x10;
+
+const V2_HEADER_SIZE: usize = 40;
+
+/// Errors that occur during serialization.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum SerializeError {
+    /// A count above i64::max_value() cannot be zig-zag encoded, and therefore cannot be
+    /// serialized.
+    CountNotSerializable,
+    /// Internal calculations cannot be represented in `usize`. Use smaller histograms or beefier
+    /// hardware.
+    UsizeTypeTooSmall,
+    /// An i/o operation failed.
+    IoError
+}
+
+impl std::convert::From<std::io::Error> for SerializeError {
+    fn from(_: std::io::Error) -> Self {
+        SerializeError::IoError
+    }
+}
+
+/// Serializer for the V2 binary format.
+/// Serializers are intended to be re-used for many histograms.
```

Comment:
Interesting... I was thinking of having a per-version trait like:
```rust
pub trait V2Serialize {
    fn to_bytes(&self) -> Result>Vec<u8>>;
    fn serialize_with(&self, &mut Vec<u8>) -> Result<()>;
}
```
Now, we could implement `V2Serialize`, `V2CompressedSerialize`, etc. for `Histogram`, and users can choose which one they want depending on which trait they import:
```rust
fn main() {
    let h: Histogram<_> = ...;
    {
        use hdrsample::serialize::V2Serialize;
        h.to_bytes(); // will use V2
    }
    {
        use hdrsample::serialize::V2CompressedSerialize;
        h.to_bytes(); // will use V2 compressed
    }
}
```
This gives a consistent API across the different versions, and there's no need for keeping this extra struct around. It is also ergonomic (if you only ever want `V2Serialize`, you just `use` that at the top of your file), and fits well within idiomatic Rust.

I agree with you that people should think about serialization, and I think this API caters to that, but it is *also* easy to use for those who *don't* have a use-case quite as critical as yours.